src/blackbox/reader_frame.go

```diff
@@ -90,95 +90,102 @@ func NewFrameReader(dec *stream.Decoder, frameDef LogDefinition, opts *FrameRead
+func (f *FrameReader) ReadNextFrame() Frame {
+	// create some base frame to make sure we always have something to return
+	// even if we could not read any frame data
+	var frame Frame = &baseFrame{}
+
+	// ---------------------------- PRE-READING ----------------------------- //
+
 	start := f.dec.BytesRead()
+	frameType, err := f.dec.ReadByte()
+	frame.setError(err)
 
-	command, err := f.dec.ReadByte()
-	if err != nil {
-		return nil, err
+	// if frame had any errors on pre-reading stage - count as corrupted and return
+	if frame.Error() != nil {
+		f.countFrameAsCorrupted(frame)
+		return frame
 	}
-	frameType := LogFrameType(command)
 
-	//Next byte should be readable if stream is valid
-	switch command {
+	// -------------------------- READING/PARSING --------------------------- //
+
+	// parse next frame data based on its type
+	switch frameType {
 	case LogFrameEvent:
 		eventType, eventValues, err := f.parseEventFrame(f.dec)
-		if err != nil {
-			return nil, err
-		}
-		end := f.dec.BytesRead()
-		frame := NewEventFrame(*eventType, eventValues, start, end)
-		_, err = f.PreComplete(frame)
-		return frame, err
+		frame = NewEventFrame(eventType, eventValues, start, f.dec.BytesRead())
+		frame.setError(err)
 
 	case LogFrameIntra:
 		values, err := ParseFrame(f.frameDef, f.frameDef.FieldsI, f.Previous, f.PreviousPrevious, f.dec, f.raw, f.LastSkippedFrames)
-		if err != nil {
-			return nil, err
-		}
-		end := f.dec.BytesRead()
-		frame := NewMainFrame(frameType, values, start, end)
-		_, err = f.PreComplete(frame)
-		return frame, err
+		frame = NewMainFrame(frameType, values, start, f.dec.BytesRead())
+		frame.setError(err)
 
 	case LogFrameInter:
 		f.LastSkippedFrames = f.countIntentionallySkippedFrames()
 		values, err := ParseFrame(f.frameDef, f.frameDef.FieldsP, f.Previous, f.PreviousPrevious, f.dec, f.raw, f.LastSkippedFrames)
-		if err != nil {
-			return nil, err
-		}
-		end := f.dec.BytesRead()
-		frame := NewMainFrame(frameType, values, start, end)
-		_, err = f.PreComplete(frame)
-		return frame, err
+		frame = NewMainFrame(frameType, values, start, f.dec.BytesRead())
+		frame.setError(err)
 
 	case LogFrameSlow:
 		values, err := ParseFrame(f.frameDef, f.frameDef.FieldsS, nil, nil, f.dec, f.raw, 0)
-		if err != nil {
-			return nil, err
-		}
-		end := f.dec.BytesRead()
-		frame := NewSlowFrame(values, start, end)
-		_, err = f.PreComplete(frame)
-		return frame, err
+		frame = NewSlowFrame(values, start, f.dec.BytesRead())
+		frame.setError(err)
 
 	default:
+		frame.setError(frameErrorUnsupportedType(frameType))
 		//TODO: If a P frame is corrupt here, we should optionally drop the previous one (As the original implementation does)
 		f.flightLogReaderInvalidateStream()
-		f.Stats.TotalCorruptedFrames++
-		return nil, errors.Errorf("Frame type '%s' (b'%b') is not supported", string(command), command)
 	}
+
+	// if frame had any errors on reading/parsing stage - count as corrupted and return
```

Comment:
I just had a look. You can actually drop those two `if frame.Error() != nil` and later improve the `PreComplete` method to handle that case